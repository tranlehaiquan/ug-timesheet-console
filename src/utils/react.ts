import * as React from 'react'

type IdentityFunction = <T>(fn: T) => T

export const typedMemo: IdentityFunction = React.memo

export const getDisplayName = <T extends any = any>(WrappedComponent: React.ComponentType<T>): string => {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}
