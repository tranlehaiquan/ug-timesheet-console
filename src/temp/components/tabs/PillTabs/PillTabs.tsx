import * as React from 'react'
import classNames from 'classnames'

export interface PillTab {
  id: string
  name: string
}

export interface IPillTabsProps extends React.HTMLAttributes<HTMLUListElement> {
  /**
   * Tabs to render
   */
  tabs: PillTab[]
  /**
   * On tab click function, returns the ID of the tab that was clicked
   */
  onTabClick: (tabClicked: string) => void
  /**
   * By default the component will use the first tab as the default active tab.
   * If this needs to be different, you can pass this prop through.
   * It should be used for routing tabs so the components state will be in sync
   * across page rereshes
   */
  initialActiveTab?: PillTab
  /**
   * Class name to override the tab parent. HTMLElement = <ul>
   */
  className?: string
  /**
   * Class names to override the individual tab. HTMLElement = <li>
   */
  tabClassName?: string
  /**
   * Class names to override the active styles of a tab. HTMLElement = <li>
   */
  activeTabClassName?: string
}

export interface IPillTabsState {
  activeTab: PillTab['id']
}

export class PillTabs extends React.PureComponent<IPillTabsProps, IPillTabsState> {
  state = {
    activeTab: this.props.initialActiveTab ? this.props.initialActiveTab.id : this.props.tabs[0].id
  }

  handleTabClick = (tabClickedId: PillTab['id']) => () => {
    this.setState({ activeTab: tabClickedId })
    this.props.onTabClick(tabClickedId)
  }

  render() {
    const { activeTab } = this.state
    const { tabs, className, tabClassName, activeTabClassName, ...otherProps } = this.props

    const classes = classNames('sk-list-reset sk-flex sk-bg-navy-lighter sk-p-2 sk-text-xs', className)
    const activeClasses = 'sk-bg-navy-light sk-opacity-100 sk-shadow'
    const tabClasses = 'sk-mr-4 sk-px-3 sk-py-1 sk-uppercase sk-rounded sk-text-white sk-cursor-pointer sk-font-semibold sk-tracking-wider sk-opacity-75 hover:sk-opacity-100 hover:sk-bg-navy-light hover:sk-shadow-tabs'

    return (
      <ul { ...otherProps } className={ classes }>
        { tabs.map(tab => {
          const combinedTabClasses = classNames(tabClasses, tabClassName, { [`${activeClasses} ${activeTabClassName ? activeTabClassName : ''}`]: activeTab === tab.id })

          return (
            <li className={ combinedTabClasses } onClick={ this.handleTabClick(tab.id) } key={ tab.id }>
              { tab.name }
            </li>
          )
        }) }
      </ul>
    )
  }
}
