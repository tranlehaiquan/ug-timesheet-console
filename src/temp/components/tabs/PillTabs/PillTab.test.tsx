import * as React from "react";
import { render, shallow } from "enzyme";

import { PillTabs, IPillTabsState } from "./PillTabs";

const tabs = [
  { name: "Tab one", id: "tab-one" },
  { name: "Tab two", id: "tab-two" },
  { name: "Tab three", id: "tab-three" },
  { name: "Tab four", id: "tab-four" },
];

describe("PillTabs", () => {
  test("it renders correctly", () => {
    // Arrange
    const component = render(<PillTabs tabs={tabs} onTabClick={jest.fn()} />);

    // Assert
    expect(component).toMatchSnapshot();
  });

  test("renders with correct initial state", () => {
    // Arrange
    const component = shallow<{}, IPillTabsState>(
      <PillTabs tabs={tabs} onTabClick={jest.fn()} />
    );

    // Assert
    expect(component.state().activeTab).toBe(tabs[0].id);
  });

  test("renders with correct state when passed a initialActiveTab", () => {
    // Arrange
    const component = shallow<{}, IPillTabsState>(
      <PillTabs tabs={tabs} initialActiveTab={tabs[2]} onTabClick={jest.fn()} />
    );

    // Assert
    expect(component.state().activeTab).toBe(tabs[2].id);
  });

  test("clicking a new tab sets the correct state", () => {
    // Arrange
    const onTabClickSpy = jest.fn();
    const component = shallow<{}, IPillTabsState>(
      <PillTabs tabs={tabs} onTabClick={onTabClickSpy} />
    );

    // Act
    expect(component.state().activeTab).toBe("tab-one");
    const tabThree = component
      .find("li")
      .findWhere((item) => item.text() === "Tab three")
      .first();
    tabThree.simulate("click");

    // Assert
    expect(component.state().activeTab).toBe("tab-three");
    expect(onTabClickSpy).toBeCalled;
    expect(onTabClickSpy).toBeCalledWith("tab-three");
  });
});
