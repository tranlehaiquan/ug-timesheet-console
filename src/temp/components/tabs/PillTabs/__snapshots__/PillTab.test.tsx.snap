// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PillTabs it renders correctly 1`] = `
<ul
  class="sk-list-reset sk-flex sk-bg-navy-lighter sk-p-2 sk-text-xs"
>
  <li
    class="sk-mr-4 sk-px-3 sk-py-1 sk-uppercase sk-rounded sk-text-white sk-cursor-pointer sk-font-semibold sk-tracking-wider sk-opacity-75 hover:sk-opacity-100 hover:sk-bg-navy-light hover:sk-shadow-tabs sk-bg-navy-light sk-opacity-100 sk-shadow "
  >
    Tab one
  </li>
  <li
    class="sk-mr-4 sk-px-3 sk-py-1 sk-uppercase sk-rounded sk-text-white sk-cursor-pointer sk-font-semibold sk-tracking-wider sk-opacity-75 hover:sk-opacity-100 hover:sk-bg-navy-light hover:sk-shadow-tabs"
  >
    Tab two
  </li>
  <li
    class="sk-mr-4 sk-px-3 sk-py-1 sk-uppercase sk-rounded sk-text-white sk-cursor-pointer sk-font-semibold sk-tracking-wider sk-opacity-75 hover:sk-opacity-100 hover:sk-bg-navy-light hover:sk-shadow-tabs"
  >
    Tab three
  </li>
  <li
    class="sk-mr-4 sk-px-3 sk-py-1 sk-uppercase sk-rounded sk-text-white sk-cursor-pointer sk-font-semibold sk-tracking-wider sk-opacity-75 hover:sk-opacity-100 hover:sk-bg-navy-light hover:sk-shadow-tabs"
  >
    Tab four
  </li>
</ul>
`;
